### PR TITLE
SCP-1687: use freer logging in Cardano.Node.Server

### PR DESCRIPTION
**Pull request by gilligan**
**TL;DR**: Update Node Mock Server to use the tracer from iohk-monitoring

- Introduce Cardano.Node.Types.MockServerLogMsg
- Refactor Cardano.Node.Server.main

----

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `nix-shell shell.nix --run updateMaterialized` to update the materialized Nix files
    - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
    - [ ] `nix-shell shell.nix --run fix-stylish-haskell` to fix any formatting issues
- If you changed any Purescript files:
    - [ ] `nix-shell shell.nix --run fix-purty` to fix any formatting issues

Pre-merge checklist:
- [ ] Someone approved it
- [ ] Commits have useful messages
- [ ] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

**Comment by gilligan** [comment]
This will need #2724 to pass CI

File: plutus-pab/logging.yaml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+rotation: null
```
Comment (gilligan) [inline plutus-pab/logging.yaml:1]: oops - this file should not have been committed

File: plutus-pab/src/Cardano/Node/Server.hs [hunk 2]
```diff
@@ -43,46 +44,43 @@ app serverHandler clientHandler stateVar =
...
+          (newFollower :<|> getBlocks))
+
+data Ctx = Ctx { serverHandler :: Server.ServerHandler
+               , clientHandler :: Client.ClientHandler
+               , serverState   ::  MVar AppState
+               , clientState   :: MVar AppState
+               }
+
+main :: Trace IO MockServerLogMsg -> MockServerConfig -> Availability -> IO ()
+main trace MockServerConfig {..} availability = runLogEffects trace $ do
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Node/Server.hs:56]: I'd prefer to name the fields of `MockServerConfig` explicitly, without `RecordWildCards`. Makes it easier to read IMO.

File: plutus-pab/src/Cardano/Node/Mock.hs [hunk 3]
```diff
@@ -137,26 +111,27 @@ runChainEffects ::
...
+        $ interpret (handleZoomedState chainState)
+        $ interpret (handleZoomedState followerState)
+        $ interpret (mapLog ProcessingChainEvent)
+        $ reinterpret CE.handleChain
+        $ interpret (mapLog ProcessingChainEvent)
+        $ reinterpret Chain.handleControlChain
+        $ interpret (mapLog FollowerMsg)
         $ FE.handleNodeFollower
-        $ interpret (mapLog NodeGenRandomTxMsg)
+        $ subsume
```
Comment (gilligan) [inline plutus-pab/src/Cardano/Node/Mock.hs:134]: @j-mueller I *think* `subsume` should be right here? type tetris first led me to something like

```
$ interpret $ (mapLog id)
```
which obviously seemed rather silly :joy:

File: plutus-pab/src/Cardano/Node/Types.hs [hunk 4]
```diff
@@ -46,19 +81,113 @@ data MockServerConfig =
...
+        RemovingOldBlocks         ->  mkObjectStr "Starting block reaper thread (old blocks will be removed)" ()
+        StartingMockServer p      ->  mkObjectStr "Starting Mock Node Server on port " (Tagged @"port" p)
+        StartingSlotCoordination  ->  mkObjectStr "" ()
+        ProcessingChainEvent e    ->  mkObjectStr "Processing chain event" (Tagged @"event" e)
+        BlockOperation e          ->  mkObjectStr "Block operation" (Tagged @"event" e)
+        CreatingRandomTransaction ->  mkObjectStr "Creating random transaction" ()
+        FollowerMsg m             -> toObject v m
+
+-- | logging type for 'Cardano.Node.Follower'
+data NodeFollowerLogMsg =
```
Comment (gilligan) [inline plutus-pab/src/Cardano/Node/Types.hs:137]: @j-mueller Maybe it would be nice to have some very single sentence descriptions for each of these *?* also for me, as i really don't actually know what they refer to/mean and i didn't want to write something too redundant.

File: plutus-pab/src/Cardano/Node/Types.hs [hunk 5]
```diff
@@ -70,17 +199,16 @@ initialAppState wallets =
...
-    deriving newtype (ToJSON, FromJSON, ToHttpApiData, FromHttpApiData, Integral, Enum, Real, Num, Pretty)
+-- | 'ChainState' with initial values
+initialChainState :: Trace.InitialDistribution -> ChainState
+initialChainState =
+    view EM.chainState .
+    MultiAgent.emulatorStateInitialDist . Map.mapKeys EM.walletPubKey
 
-makeLenses 'AppState
+-- | Isomorphism to go between 'NodeFollowerState' and 'Map FollowerID Int'
+_NodeFollowerState :: Iso' NodeFollowerState (Map FollowerID Int)
```
Comment (gilligan) [inline plutus-pab/src/Cardano/Node/Types.hs:213]: Does this adhere to a naming convention of some sort? `_NodeFollowreState` seems odd to me for this iso but i am maybe just not aware..

File: plutus-pab/src/Cardano/Node/Mock.hs (hunk 3 above)
Comment (j-mueller) [reply plutus-pab/src/Cardano/Node/Mock.hs:134]: Yes `subsume` is right 👍

File: plutus-pab/src/Cardano/Node/Server.hs [hunk 6]
```diff
@@ -11,78 +11,80 @@ module Cardano.Node.Server
...
         (healthcheck :<|> addTx :<|> getCurrentSlot :<|>
          (genRandomTx :<|>
           consumeEventHistory stateVar) :<|>
-          (newFollower :<|> getBlocks)
-          )
+          (newFollower :<|> getBlocks))
+
+data Ctx = Ctx { serverHandler :: Server.ServerHandler
+               , clientHandler :: Client.ClientHandler
+               , serverState   ::  MVar AppState
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Node/Server.hs:51]:
```suggestion
               , serverState   :: MVar AppState
```

File: plutus-pab/src/Cardano/Node/Types.hs (hunk 4 above)
Comment (j-mueller) [reply plutus-pab/src/Cardano/Node/Types.hs:137]: In prinicple yes, but the `NodeFollower` mechanism is going to be removed and replaced by the the socket protocol from `ouroboros-network` (by @raduom). So we can leave it as it is.

The idea was that each node follower periodically asks the node for new blocks since the last time they've asked. But with the socket protocol this polling is going to be replaced by the node pushing new blocks to the clients.

File: plutus-pab/src/Cardano/Node/Types.hs (hunk 5 above)
Comment (j-mueller) [reply plutus-pab/src/Cardano/Node/Types.hs:213]: Yes the name is correct but you could use [`makePrisms`](http://hackage.haskell.org/package/lens-4.15.4/docs/Control-Lens-TH.html#v:makePrisms) which generates isos when possible.

File: plutus-pab/src/Cardano/Node/Types.hs [hunk 7]
```diff
@@ -46,19 +81,113 @@ data MockServerConfig =
...
+instance Pretty BlockEvent where
+    pretty = \case
+        NewSlot          -> "Adding a new slot"
+        NewTransaction t -> "Adding a transaction " <+> pretty (Ledger.txId t)
+
+
+-- State --------------------------------------------------------------------------------------------------------------
+
+
+newtype NodeFollowerState = NodeFollowerState { _unNodeFollowerState :: Map FollowerID Int }
```
Comment (j-mueller) [inline plutus-pab/src/Cardano/Node/Types.hs:178]:
```suggestion
newtype NodeFollowerState = NodeFollowerState { unNodeFollowerState :: Map FollowerID Int }
```

generally we use `un` for newtypes unless something like `run` makes more sense.

**Review by j-mueller — APPROVED** [review]

File: plutus-pab/src/Cardano/Node/Types.hs (hunk 4 above)
Comment (raduom) [reply plutus-pab/src/Cardano/Node/Types.hs:137]: I did that, but there will be issues with the semantics we give to the follower messages in tests. So what I think I will do, finally, is change the follower algebra to be more suited (stateful) to the way the socket protocol works and implement something sensible for testing, though I am not sure what exactly that would be.

Comment (gilligan) [reply plutus-pab/src/Cardano/Node/Types.hs:137]: I will leave it as is for now, but I would appreciate some comments somewhere that at least broadly outline what a "node follower" is. Either in the Types module or in the Follower module itself or just _here_ if need be and i can add it somewhere later on my own.

File: plutus-pab/src/Cardano/Node/Types.hs (hunk 7 above)
Comment (gilligan) [reply plutus-pab/src/Cardano/Node/Types.hs:178]: I didn't write that code i only moved it around in the file :sweat_smile:  But makes sense to me :+1:
